### PR TITLE
Added gatk addorreplacereadgroups feature

### LINKED ISSUE #838: new module: gatk4/addorreplacereadgroups
I think it would be good to have a module for [gatk4](https://bioconda.github.io/recipes/gatk4/README.html)/addorreplacereadgroups 

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by simeonhebrew**
## PR checklist

Closes #838 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+
```
Comment (maxulysse) [inline modules/gatk4/addorreplacereadgroups/main.nf:4]:
```suggestion

```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+options        = initOptions(params.options)
+
+process GATK4_ADDORREPLACEREADGROUPS {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+
```
Comment (maxulysse) [inline modules/gatk4/addorreplacereadgroups/main.nf:15]:
```suggestion

```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.3.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/gatk4:4.2.3.0--hdfd78af_0"
+    } else {
+        container "quay.io/biocontainers/gatk4:4.2.3.0--hdfd78af_0"
+    }
```
Comment (maxulysse) [inline modules/gatk4/addorreplacereadgroups/main.nf:17]: Sorry, included an extra `}` in my suggestions
```suggestion
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+params.options = [:]
+options        = initOptions(params.options)
+
+process GATK4_ADDORREPLACEREADGROUPS {
+    tag "$meta.id"
+    label 'process_low'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.3.0" : null)
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:12]:
```suggestion
        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }

    conda (params.enable_conda ? "bioconda::gatk4=4.2.3.0" : null)
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    } else {
+        container "quay.io/biocontainers/gatk4:4.2.3.0--hdfd78af_0"
+    }
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.bam"), emit: bam
+    path "version.yml",             emit: versions
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:24]:
```suggestion
    tuple val(meta), path("*.bam"), emit: bam
    path "version.yml"            , emit: versions
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:29]:
```suggestion
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    gatk AddOrReplaceReadGroups" \\
+        -I $bam \\
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:32]:
```suggestion
    gatk 
        AddOrReplaceReadGroups \\
        -I $bam \\
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        -O ${prefix}.bam \\
+        -ID=4 \\
+        -LB=lib1 \\
+        -PL=ILLUMINA \\
+        -PU=Unit1 \\
+        -SM=20 \\
+        $options.args
+
+
+cat <<-END_VERSIONS > versions.yml
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:42]:
```suggestion
    cat <<-END_VERSIONS > versions.yml
```

File: modules/gatk4/addorreplacereadgroups/meta.yml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  - gatk4:
+      description: Developed in the Data Sciences Platform at the Broad Institute, the toolkit offers a wide variety of tools
+            with a primary focus on variant discovery and genotyping. Its powerful processing engine
+            and high-performance computing features make it capable of taking on projects of any size.
+      homepage: https://gatk.broadinstitute.org/hc/en-us
+      documentation: https://gatk.broadinstitute.org/hc/en-us/categories/360002369672s
+      tool_dev_url: https://github.com/broadinstitute/gatk
+      doi: "10.1158/1538-7445.AM2017-3590"
+      licence: ['Apache-2.0']
+
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/meta.yml:18]:
```suggestion
```

File: modules/gatk4/addorreplacereadgroups/meta.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
+name: gatk4_addorreplacereadgroups
+description:  This feature of gatk assigns all the reads in a file to a single new read-group
+keywords:
+- gatk
+- bam
+- read groups
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/meta.yml:6]:
```suggestion
  - gatk
  - bam
  - read groups
```

File: modules/gatk4/addorreplacereadgroups/meta.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  - bam:
+      type: file
+      description: BAM file with modified read groups
+      pattern: "*.{bam}"
+
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.yml}"
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/meta.yml:46]:
```suggestion
  - versions:
      type: file
      description: File containing software versions
      pattern: "*.{versions.yml}"
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    gatk AddOrReplaceReadGroups" \\
+        -I $bam \\
+        -O ${prefix}.bam \\
+        -ID=4 \\
+        -LB=lib1 \\
+        -PL=ILLUMINA \\
+        -PU=Unit1 \\
+        -SM=20 \\
+        $options.args
+
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:40]:
```suggestion
```

File: modules/gatk4/addorreplacereadgroups/main.nf [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    gatk AddOrReplaceReadGroups" \\
+        -I $bam \\
+        -O ${prefix}.bam \\
+        -ID=4 \\
+        -LB=lib1 \\
+        -PL=ILLUMINA \\
+        -PU=Unit1 \\
+        -SM=20 \\
```
Comment (drpatelh) [inline modules/gatk4/addorreplacereadgroups/main.nf:38]: These should be provided as `input` options to the module so they can be overriden. Maybe they can be passed in via the `meta` map.

Comment (drpatelh) [reply modules/gatk4/addorreplacereadgroups/main.nf:38]: Don't know how @maxulysse is passing read group info in sarek but worth seeing what he is doing there.

**Comment by drpatelh** [comment]
Still a little work to do @simeonhebrew. You generally need to add 6 files when adding a new module including tests like in [this PR](https://github.com/nf-core/modules/pull/977)

File: modules/gatk4/addorreplacereadgroups/main.nf (hunk 13 above)
Comment (simeonhebrew) [reply modules/gatk4/addorreplacereadgroups/main.nf:38]: Okay, working on it.

**Comment by JoseEspinosa** [comment]
Hi @simeonhebrew, thank you for this PR! We are merging as many modules as possible now due to and impending restructuring of the entire repo. This will mean you will need to update the module to reflect these changes before it can be merged in the future. It appears like this module isn't ready to be merged so if applicable, we are converting it to draft and adding the WIP label. If this isn't the case please let us know and we will try to get the module in before the changes. Thanks again!

**Comment by matthdsm** [comment]
Hi there!

We've noticed there hasn't been much activity here. Are you still planning on working on this?
If not, you can ignore this message and we'll close your PR in about 2 weeks

Cheers
the nf-core maintainers

**Pull request closed without merging**